File: solune/backend/src/models/cleanup.py
"""Pydantic models for repository cleanup operations."""

from typing import Any

from pydantic import BaseModel, Field

# ── Request Models ──────────────────────────────────────────────────


class CleanupPreflightRequest(BaseModel):
    """Request body for the preflight check."""

    owner: str = Field(min_length=1, description="Repository owner (user or org)")
    repo: str = Field(min_length=1, description="Repository name")
    project_id: str = Field(min_length=1, description="GitHub Projects v2 project node ID")


class IssueToDelete(BaseModel):
    """An orphaned issue to permanently delete."""

    number: int = Field(description="GitHub issue number")
    node_id: str = Field(min_length=1, description="GitHub GraphQL node ID")


class CleanupExecuteRequest(BaseModel):
    """Request body to execute the cleanup operation."""

    owner: str = Field(min_length=1, description="Repository owner")
    repo: str = Field(min_length=1, description="Repository name")
    project_id: str = Field(min_length=1, description="Project board ID (for audit trail)")
    branches_to_delete: list[str] = Field(description="Branch names to delete")
    prs_to_close: list[int] = Field(description="PR numbers to close")
    issues_to_delete: list[IssueToDelete] = Field(
        default=[],
        description="Orphaned issues to permanently delete (requires node_id)",
    )


# ── Response Models ─────────────────────────────────────────────────


class BranchInfo(BaseModel):
    """Branch details for preflight response."""

    name: str
    eligible_for_deletion: bool
    linked_issue_number: int | None = None
    linked_issue_title: str | None = None
    linking_method: str | None = None
    preservation_reason: str | None = None
    deletion_reason: str | None = None


class PullRequestInfo(BaseModel):
    """PR details for preflight response."""

    number: int
    title: str
    head_branch: str
    referenced_issues: list[int] = []
    eligible_for_deletion: bool
    preservation_reason: str | None = None
    deletion_reason: str | None = None


class OrphanedIssueInfo(BaseModel):
    """An app-created GitHub Issue that is no longer on the project board."""

    number: int
    title: str
    labels: list[str] = []
    html_url: str | None = None
    node_id: str | None = None


class IssueInfo(BaseModel):
    """An open issue surfaced in the cleanup modal (preserve section)."""

    number: int
    title: str
    labels: list[str] = []
    html_url: str | None = None
    node_id: str | None = None
    preservation_reason: str | None = None


class CleanupPreflightResponse(BaseModel):
    """Response from preflight endpoint."""

    branches_to_delete: list[BranchInfo]
    branches_to_preserve: list[BranchInfo]
    prs_to_close: list[PullRequestInfo]
    prs_to_preserve: list[PullRequestInfo]
    orphaned_issues: list[OrphanedIssueInfo] = []
    issues_to_preserve: list[IssueInfo] = []
    open_issues_on_board: int
    has_permission: bool
    permission_error: str | None = None


class CleanupItemResult(BaseModel):
    """Result of a single deletion/close operation."""

    item_type: str = Field(description="'branch', 'pr', or 'issue'")
    identifier: str = Field(description="Branch name, PR number, or issue number as string")
    action: str = Field(description="'deleted', 'closed', 'preserved', or 'failed'")
    reason: str | None = None
    error: str | None = None


class CleanupExecuteResponse(BaseModel):
    """Final summary response from execute endpoint."""

    operation_id: str
    branches_deleted: int
    branches_preserved: int
    prs_closed: int
    prs_preserved: int
    issues_deleted: int = 0
    errors: list[CleanupItemResult]
    results: list[CleanupItemResult]


# ── Database Row Model ──────────────────────────────────────────────


class CleanupAuditLogRow(BaseModel):
    """Represents a cleanup_audit_logs database row."""

    id: str
    github_user_id: str
    owner: str
    repo: str
    project_id: str
    started_at: str
    completed_at: str | None
    status: str
    branches_deleted: int
    branches_preserved: int
    prs_closed: int
    prs_preserved: int
    issues_closed: int = 0
    errors_count: int
    details: dict[str, Any] | None  # Parsed JSON object (stored as TEXT in SQLite)


# ── History Response ────────────────────────────────────────────────


class CleanupHistoryResponse(BaseModel):
    """Response for the audit history endpoint."""

    operations: list[CleanupAuditLogRow]
    count: int

File: solune/frontend/src/types/cleanup.ts
/**
 * Cleanup domain types.
 */

export interface BranchInfo {
  name: string;
  eligible_for_deletion: boolean;
  linked_issue_number: number | null;
  linked_issue_title: string | null;
  linking_method: string | null;
  preservation_reason: string | null;
  deletion_reason: string | null;
}

export interface PullRequestInfo {
  number: number;
  title: string;
  head_branch: string;
  referenced_issues: number[];
  eligible_for_deletion: boolean;
  preservation_reason: string | null;
  deletion_reason: string | null;
}

export interface OrphanedIssueInfo {
  number: number;
  title: string;
  labels: string[];
  html_url: string | null;
  node_id: string | null;
}

export interface IssueInfo {
  number: number;
  title: string;
  labels: string[];
  html_url: string | null;
  node_id: string | null;
  preservation_reason: string | null;
}

export interface IssueToDelete {
  number: number;
  node_id: string;
}

export interface CleanupPreflightResponse {
  branches_to_delete: BranchInfo[];
  branches_to_preserve: BranchInfo[];
  prs_to_close: PullRequestInfo[];
  prs_to_preserve: PullRequestInfo[];
  orphaned_issues: OrphanedIssueInfo[];
  issues_to_preserve: IssueInfo[];
  open_issues_on_board: number;
  has_permission: boolean;
  permission_error: string | null;
}

export interface CleanupItemResult {
  item_type: 'branch' | 'pr' | 'issue';
  identifier: string;
  action: 'deleted' | 'closed' | 'preserved' | 'failed';
  reason: string | null;
  error: string | null;
}

export interface CleanupExecuteRequest {
  owner: string;
  repo: string;
  project_id: string;
  branches_to_delete: string[];
  prs_to_close: number[];
  issues_to_delete: IssueToDelete[];
}

/** Payload from the confirm modal — the user's final selections after toggling. */
export interface CleanupConfirmPayload {
  branches_to_delete: string[];
  prs_to_close: number[];
  issues_to_delete: IssueToDelete[];
}

export interface CleanupExecuteResponse {
  operation_id: string;
  branches_deleted: number;
  branches_preserved: number;
  prs_closed: number;
  prs_preserved: number;
  issues_deleted: number;
  errors: CleanupItemResult[];
  results: CleanupItemResult[];
}

export interface CleanupAuditLogEntry {
  id: string;
  started_at: string;
  completed_at: string | null;
  status: string;
  branches_deleted: number;
  branches_preserved: number;
  prs_closed: number;
  prs_preserved: number;
  errors_count: number;
  details: {
    results: CleanupItemResult[];
  } | null;
}

export interface CleanupHistoryResponse {
  operations: CleanupAuditLogEntry[];
  count: number;
}
